In [1]:
# import the necessary libraries
import pandas as pd

In [2]:
#load the document
df = pd.read_csv("GPTReviews.csv",header=None,on_bad_lines='skip',encoding='utf8')

C:\Users\Admin\AppData\Local\Temp\ipykernel_8304\315083873.py:2: DtypeWarning: Columns (1,5) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("GPTReviews.csv",header=None,on_bad_lines='skip',encoding='utf8')


In [3]:
df

0       1  \
0                               Name  Rating   
1                          Musa Maan       4   
2                              Lorax       5   
3                       Ramesh Singh       5   
4                      Gurjaap Singh       3   
...                              ...     ...   
99996   Martín Aurelio Mamani aviles       5   
99997                       Arko Roy       5   
99998                    Aniket Paul       5   
99999                      Guru Guru       4   
100000                 Andy Dressler       3   

                                                        2  \
0                                                 Comment   
1       pictures information is not good but overall best   
2                                             Very useful   
3       This app is very good for students who are str...   
4                                          limited chats😤   
...                                                   ...   
99996                                         para laptop   
99997   It is a wonderful app.It really help me in my ...   
99998                                Most batter good app   
99999   I think it is very good app. it help me alot b...   
100000  I didnt find using this app at all special or ...   

                          3        4          5  \
0                      Date  Country  Thumbs Up   
1       2024-07-18 18:19:22       us          0   
2       2024-07-18 18:18:15       us          0   
3       2024-07-18 18:18:00       us          0   
4       2024-07-18 18:15:15       us          0   
...                     ...      ...        ...   
99996   2023-11-21 03:43:46       us          0   
99997   2023-11-21 03:42:07       us          0   
99998   2023-11-21 03:36:53       us          0   
99999   2023-11-21 03:35:24       us          0   
100000  2023-11-21 03:13:27       us          0   

                                           6            7  
0                                  Review ID  App Version  
1       4312ea55-28b5-4abd-b007-c0c4909c25bc   1.2024.192  
2       249e38e0-9d3a-4392-896d-88fd7e4499f6          NaN  
3       d1ad5cbb-29c1-468b-8e00-31d82f4ee943   1.2024.192  
4       6142e817-ac9a-4beb-aa8b-6ed8a273ed82   1.2024.157  
...                                      ...          ...  
99996   6b973659-26c4-40e6-8265-541a80af6f6f   1.2023.313  
99997   609c1cd2-815c-40d8-b5f9-352084f4f67f   1.2023.313  
99998   afa94f1c-7193-48a9-a012-b7f49d8c0844   1.2023.313  
99999   d50fe21d-5bb3-427d-9b2c-71cbf4866e5f   1.2023.313  
100000  4a04ce7e-94d3-42d7-bde8-c8e63acc1ffc   1.2023.313  

[100001 rows x 8 columns]

In [4]:
#retrieve the 1st 100 rows
df=df.iloc[0:100]

In [5]:
# drpping the unwanted columns
cols_to_del=[1,4,5,6,7]
df=df.drop(cols_to_del,axis=1)
df=df.drop(0,axis=0)

In [7]:
#renaming the columns
df.columns=["Name","Comment","Date"]

In [8]:
date=df['Date'].str.split(' ',n=1,expand=True)
df["Date"]=date[0]
df.dropna(inplace=True)

In [13]:
df

Name                                            Comment  \
1         Musa Maan  pictures information is not good but overall best   
2             Lorax                                        Very useful   
3      Ramesh Singh  This app is very good for students who are str...   
4     Gurjaap Singh                                     limited chats😤   
5       Riya Sharma                  i like this app very muchh ❤️❤️❤️   
..              ...                                                ...   
95            Reece  In the corner of my phone screen it said somet...   
96  Jullan Maglinte  Thank to ChatGPT, I am very Happy to learn mor...   
97   Yogmaya Sharma                               Best app , Love it .   
98    Vishal Nimase       This is app is very good and useful to study   
99  Sariful Mallick  My very favorite app which I try to find in an...   

          Date  
1   2024-07-18  
2   2024-07-18  
3   2024-07-18  
4   2024-07-18  
5   2024-07-18  
..         ...  
95  2024-07-18  
96  2024-07-18  
97  2024-07-18  
98  2024-07-18  
99  2024-07-18  

[99 rows x 3 columns]

In [9]:
#extracting features using Hashing Vectorizer
from sklearn.feature_extraction.text import HashingVectorizer
vectorizer = HashingVectorizer(n_features=1000, stop_words='english',alternate_sign=False)
hashed_features = vectorizer.fit_transform(df['Comment'])

In [10]:
print(hashed_features)

  (0, 187)	0.4472135954999579
  (0, 729)	0.4472135954999579
  (0, 831)	0.4472135954999579
  (0, 954)	0.4472135954999579
  (0, 969)	0.4472135954999579
  (1, 742)	1.0
  (2, 15)	0.2886751345948129
  (2, 106)	0.2886751345948129
  (2, 249)	0.2886751345948129
  (2, 436)	0.2886751345948129
  (2, 729)	0.2886751345948129
  (2, 770)	0.2886751345948129
  (2, 815)	0.5773502691896258
  (2, 843)	0.2886751345948129
  (2, 940)	0.2886751345948129
  (3, 759)	0.7071067811865475
  (3, 921)	0.7071067811865475
  (4, 627)	0.5773502691896258
  (4, 815)	0.5773502691896258
  (4, 855)	0.5773502691896258
  (5, 70)	0.17677669529663687
  (5, 276)	0.17677669529663687
  (5, 314)	0.35355339059327373
  (5, 388)	0.17677669529663687
  (5, 729)	0.35355339059327373
  :	:
  (95, 466)	0.5
  (96, 201)	0.5773502691896258
  (96, 815)	0.5773502691896258
  (96, 831)	0.5773502691896258
  (97, 158)	0.5
  (97, 729)	0.5
  (97, 742)	0.5
  (97, 815)	0.5
  (98, 15)	0.20851441405707477
  (98, 74)	0.20851441405707477
  (98, 108)	0.2085144

## 456 stored elements: The total number of non-zero elements in the matrix.
## Shape (99, 500): The matrix has 99 rows and 500 columns.
## The pairs (row, column) are the coordinates of the non-zero values in the matrix.
## (0, 187) 0.4472135954999579: At row 0 and column 187, the value is approximately 0.4472.
## (1, 242) 1.0: At row 1 and column 242, the value is 1.0.

In [11]:
# LDA Model
from sklearn.decomposition import LatentDirichletAllocation
lda = LatentDirichletAllocation(n_components=5, random_state=0)
lda_model=lda.fit_transform(hashed_features)

## Each row in the output represents a document, and each column represents a topic.
## First Row [0.06434561, 0.06212836, 0.74679681, 0.06231034, 0.06441888]:
This represents the topic distribution for the first document.
The document has approximately a 74.68% probability of belonging to the third topic, with much smaller probabilities for the other topics.

In [12]:
# The components_ matrix gives insight into which words are most associated with each topic.
lda.components_

array([[0.2       , 0.2       , 0.2       , ..., 0.2       , 0.2       ,
        0.20001253],
       [0.2       , 0.2       , 0.2       , ..., 0.2       , 0.2       ,
        0.20001361],
       [0.2       , 0.2       , 0.2       , ..., 0.2       , 0.2       ,
        0.20001475],
       [0.2       , 0.2       , 0.2       , ..., 0.2       , 0.2       ,
        0.2000133 ],
       [0.2       , 0.2       , 0.2       , ..., 0.2       , 0.2       ,
        0.61697464]])

In [14]:
#the topic with highest probability is returned
df.loc[:,"Topic"]=lda_model.argmax(axis=1)

In [15]:
df

Name                                            Comment  \
1         Musa Maan  pictures information is not good but overall best   
2             Lorax                                        Very useful   
3      Ramesh Singh  This app is very good for students who are str...   
4     Gurjaap Singh                                     limited chats😤   
5       Riya Sharma                  i like this app very muchh ❤️❤️❤️   
..              ...                                                ...   
95            Reece  In the corner of my phone screen it said somet...   
96  Jullan Maglinte  Thank to ChatGPT, I am very Happy to learn mor...   
97   Yogmaya Sharma                               Best app , Love it .   
98    Vishal Nimase       This is app is very good and useful to study   
99  Sariful Mallick  My very favorite app which I try to find in an...   

          Date  Topic  
1   2024-07-18      0  
2   2024-07-18      3  
3   2024-07-18      4  
4   2024-07-18      1  
5   2024-07-18      4  
..         ...    ...  
95  2024-07-18      4  
96  2024-07-18      1  
97  2024-07-18      0  
98  2024-07-18      3  
99  2024-07-18      4  

[99 rows x 4 columns]

In [16]:
df["Topic"].value_counts()

Topic
0    31
3    23
4    21
1    16
2     8
Name: count, dtype: int64

# COUNT VECTORIZER

In [17]:
#renaming the dataset
dataset=df

In [18]:
dataset

Name                                            Comment  \
1         Musa Maan  pictures information is not good but overall best   
2             Lorax                                        Very useful   
3      Ramesh Singh  This app is very good for students who are str...   
4     Gurjaap Singh                                     limited chats😤   
5       Riya Sharma                  i like this app very muchh ❤️❤️❤️   
..              ...                                                ...   
95            Reece  In the corner of my phone screen it said somet...   
96  Jullan Maglinte  Thank to ChatGPT, I am very Happy to learn mor...   
97   Yogmaya Sharma                               Best app , Love it .   
98    Vishal Nimase       This is app is very good and useful to study   
99  Sariful Mallick  My very favorite app which I try to find in an...   

          Date  Topic  
1   2024-07-18      0  
2   2024-07-18      3  
3   2024-07-18      4  
4   2024-07-18      1  
5   2024-07-18      4  
..         ...    ...  
95  2024-07-18      4  
96  2024-07-18      1  
97  2024-07-18      0  
98  2024-07-18      3  
99  2024-07-18      4  

[99 rows x 4 columns]

In [19]:
#extracting features using Count Vectorizer
from sklearn.feature_extraction.text import CountVectorizer
count_vector=CountVectorizer(stop_words='english')
lda_count=count_vector.fit_transform(dataset['Comment'])
print(lda_count)

  (0, 173)	1
  (0, 117)	1
  (0, 93)	1
  (0, 167)	1
  (0, 28)	1
  (1, 253)	1
  (2, 93)	1
  (2, 14)	2
  (2, 218)	1
  (2, 216)	1
  (2, 219)	1
  (2, 187)	1
  (2, 248)	1
  (2, 27)	1
  (2, 104)	1
  (3, 139)	1
  (3, 37)	1
  (4, 14)	1
  (4, 137)	1
  (4, 156)	1
  (5, 93)	2
  (5, 14)	4
  (5, 90)	2
  (5, 11)	2
  (5, 247)	1
  :	:
  (95, 133)	1
  (96, 28)	1
  (96, 14)	1
  (96, 148)	1
  (97, 93)	1
  (97, 253)	1
  (97, 14)	1
  (97, 220)	1
  (98, 14)	1
  (98, 248)	1
  (98, 11)	2
  (98, 34)	1
  (98, 96)	1
  (98, 59)	1
  (98, 208)	1
  (98, 82)	1
  (98, 66)	1
  (98, 198)	1
  (98, 175)	1
  (98, 97)	1
  (98, 121)	1
  (98, 61)	1
  (98, 228)	2
  (98, 109)	1
  (98, 224)	1


In [20]:
#feature names from the document
count_vector.get_feature_names_out()

array(['15', 'able', 'account', 'achcha', 'add', 'adds', 'advancing',
       'aged', 'ai', 'allowing', 'amazing', 'answer', 'answers', 'anthor',
       'app', 'artificial', 'assistant', 'asss', 'audios', 'available',
       'awaresome', 'awesome', 'background', 'bad', 'bahut', 'based',
       'bcz', 'believe', 'best', 'better', 'bot', 'boy', 'braking',
       'capable', 'chat', 'chatbots', 'chatgpt', 'chats', 'circling',
       'click', 'close', 'clue', 'competitive', 'computer', 'conceptual',
       'contain', 'content', 'corner', 'crashing', 'creating', 'cut',
       'daily', 'definitely', 'developers', 'developing', 'development',
       'dictating', 'didn', 'dint', 'doesn', 'don', 'easily', 'easy',
       'education', 'educational', 'effortlessly', 'engine', 'enjoy',
       'especially', 'essays', 'evan', 'excellent', 'exit', 'exllent',
       'expensive', 'experience', 'expert', 'explore', 'extremely',
       'facility', 'fantastic', 'far', 'favorite', 'fine', 'finish',
       'fi

In [21]:
#total no. of feature names extracted
len(count_vector.get_feature_names_out())

275

In [22]:
# array format of the document after counter vector
print(lda_count.toarray())

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [23]:
#Applying the LDA model
from sklearn.decomposition import LatentDirichletAllocation
lda2 = LatentDirichletAllocation(n_components=5, random_state=0)
lda_count_model=lda2.fit_transform(lda_count)

In [30]:
#printing the clustered words inside the 5 topics
for index,Topic in enumerate(lda2.components_):
    results=([count_vector.get_feature_names_out()[i] for i in Topic.argsort()[-10:]])
    print(results)

['15', 'talking', 'long', 'work', 'like', 'answer', 'best', 'excellent', 'love', 'app']
['based', 'amazing', 'doesn', 'tool', 'work', 'easy', 'helpful', 'good', 'use', 'nice']
['bad', 'really', 'chat', 'great', 'information', 'friend', 'like', 'app', 'best', 'good']
['life', 'update', 'way', 'best', 'said', 'screen', 'chatgpt', 'recording', 'app', 'helpful']
['gives', 'answer', 'thank', 'need', 'nice', 'students', 'useful', 'chat', 'app', 'good']


In [31]:
Topic.shape

(275,)

In [32]:
#taking out the most likely topic for each document and assign it to the dataframe
dataset.loc[:,"Topic1"]=lda_count_model.argmax(axis=1)

In [33]:
dataset

Name                                            Comment  \
1         Musa Maan  pictures information is not good but overall best   
2             Lorax                                        Very useful   
3      Ramesh Singh  This app is very good for students who are str...   
4     Gurjaap Singh                                     limited chats😤   
5       Riya Sharma                  i like this app very muchh ❤️❤️❤️   
..              ...                                                ...   
95            Reece  In the corner of my phone screen it said somet...   
96  Jullan Maglinte  Thank to ChatGPT, I am very Happy to learn mor...   
97   Yogmaya Sharma                               Best app , Love it .   
98    Vishal Nimase       This is app is very good and useful to study   
99  Sariful Mallick  My very favorite app which I try to find in an...   

          Date  Topic  Topic1  
1   2024-07-18      0       2  
2   2024-07-18      3       4  
3   2024-07-18      4       2  
4   2024-07-18      1       4  
5   2024-07-18      4       0  
..         ...    ...     ...  
95  2024-07-18      4       3  
96  2024-07-18      1       4  
97  2024-07-18      0       0  
98  2024-07-18      3       4  
99  2024-07-18      4       0  

[99 rows x 5 columns]

In [29]:
# No. of documents in each topic
dataset['Topic1'].value_counts()

Topic1
0    30
4    25
2    18
1    13
3    13
Name: count, dtype: int64